packages/ws/src/ws/WebSocketShard.ts

```diff
@@ -543,41 +580,61 @@ export class WebSocketShard extends AsyncEventEmitter<WebSocketShardEventsMap> {
 		}
 
 		// Deal with gw wide zlib-stream compression
-		if (this.inflate) {
-			const l = decompressable.length;
+		if (this.strategy.options.compression !== null) {
 			const flush =
-				l >= 4 &&
-				decompressable[l - 4] === 0x00 &&
-				decompressable[l - 3] === 0x00 &&
-				decompressable[l - 2] === 0xff &&
-				decompressable[l - 1] === 0xff;
+				decompressable.length >= 4 &&
+				decompressable.at(-4) === 0x00 &&
+				decompressable.at(-3) === 0x00 &&
+				decompressable.at(-2) === 0xff &&
+				decompressable.at(-1) === 0xff;
 
-			const zlib = (await getZlibSync())!;
-			this.inflate.push(Buffer.from(decompressable), flush ? zlib.Z_SYNC_FLUSH : zlib.Z_NO_FLUSH);
+			if (this.nativeInflate) {
+				this.nativeInflate!.write(decompressable, 'binary');
 
-			if (this.inflate.err) {
-				this.emit(WebSocketShardEvents.Error, {
-					error: new Error(`${this.inflate.err}${this.inflate.msg ? `: ${this.inflate.msg}` : ''}`),
-				});
-			}
+				if (!flush) {
+					return null;
+				}
 
-			if (!flush) {
-				return null;
-			}
+				const [result] = await once(this.nativeInflate, 'data');
+				if (!result) {
+					return null;
+				}
 
-			const { result } = this.inflate;
-			if (!result) {
-				return null;
-			}
+				return JSON.parse(
+					typeof result === 'string' ? result : this.textDecoder.decode(result),
+				) as GatewayReceivePayload;
+			} else if (this.zLibSyncInflate) {
+				const zLibSync = (await getZLibSync())!;
+				this.zLibSyncInflate.push(Buffer.from(decompressable), flush ? zLibSync.Z_SYNC_FLUSH : zLibSync.Z_NO_FLUSH);
+
+				if (this.zLibSyncInflate.err) {
+					this.emit(WebSocketShardEvents.Error, {
+						error: new Error(
+							`${this.zLibSyncInflate.err}${this.zLibSyncInflate.msg ? `: ${this.zLibSyncInflate.msg}` : ''}`,
+						),
+					});
+				}
+
+				if (!flush) {
+					return null;
+				}
+
+				const { result } = this.zLibSyncInflate;
+				if (!result) {
+					return null;
+				}
 
-			return JSON.parse(typeof result === 'string' ? result : this.textDecoder.decode(result)) as GatewayReceivePayload;
+				return JSON.parse(
+					typeof result === 'string' ? result : this.textDecoder.decode(result),
+				) as GatewayReceivePayload;
+			}
 		}
 
 		this.debug([
 			'Received a message we were unable to decompress',
 			`isBinary: ${isBinary.toString()}`,
-			`useIdentifyCompress: ${this.useIdentifyCompress.toString()}`,
-			`inflate: ${Boolean(this.inflate).toString()}`,
+			`useIdentifyCompression: ${this.useIdentifyCompression.toString()}`,
+			`inflate: ${this.nativeInflate ? 'node:zlib' : 'zlib-sync'} (zlib-stream)`,
```

Comment:
Should make this a map too ideally, so if/when we add zstd-stream support in we dont need a ton of extra changes

Reply:
I have a different idea for this, let me know if you think this is ok.